engine/classes/Elgg/ActionsService.php

```diff
@@ -97,39 +101,52 @@ public function execute($action, $forwarder = "") {
 				break;
 			case 'logged_in':
 				if (!$user) {
-					$forward('actionloggedout');
+					return $forward('actionloggedout');
 				}
 				break;
 			default:
 				// admin or misspelling
 				if (!$user->isAdmin()) {
-					$forward('actionunauthorized');
+					return $forward('actionunauthorized');
 				}
 		}
 
 		// To quietly cancel the file, return a falsey value in the "action" hook.
 		if (!_elgg_services()->hooks->trigger('action', $action, null, true)) {
-			$forward();
+			return $forward();
 		}
 
 		$file = $this->actions[$action]['file'];
 
 		if (!is_file($file) || !is_readable($file)) {
 			$forward('actionnotfound');
 		}
+		$result = self::includeFile($file);
+		$result = _elgg_services()->hooks->trigger('action:after', $action, null, $result);
 
-		self::includeFile($file);
-		$forward();
-	}
+		if (is_array($result)) {			
+			$content = elgg_extract('content', $result);
+			$status_code = elgg_extract('status_code', $result, 200);
+			$forward_url = elgg_extract('forward_url', $result, REFERRER);
+
+			if (elgg_is_xhr() && $content) {
```

Comment:
If $status_code is in explicitly set to [301, 302, 303, 306, 307, 308] we should honor it even for XHR requests. What you think?